src/Prove.hs

```diff
@@ -0,0 +1,340 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Prove (queries) where
+
+import Control.Monad (when)
+import Data.ByteString (ByteString)
+import Data.ByteString.UTF8 (toString)
+import Data.List (intercalate, (\\), nub)
+import Data.List.NonEmpty as NonEmpty (toList)
+import Data.Map.Strict as Map (Map, lookup, fromList)
+import Data.Maybe
+
+import Data.SBV hiding (name)
+import Data.SBV.String ((.++), subStr)
+
+import RefinedAst
+import Syntax (Id, Interface(..), Decl(..))
+import Type (metaType, mkStorageBounds)
+
+-- *** Interface *** --
+
+
+{-|
+   For each Invariant claim build an SMT query consisting of:
+
+   - constants for the pre and post versions of all storage variables used by the transition system
+   - boolean predicates over the pre and post storage variables for each pass behaviour
+   - a boolean predicate for the invariant
+   - an assertion that the invariant does not hold if either of the following is true:
+      1. The constructor predicate holds
+      2. The invariant holds over the prestate and one of the method level predicates holds
+
+   If this query returns `unsat` then the invariant must hold over the transition system
+-}
+queries :: [Claim] -> [(Invariant, [Symbolic ()])]
+queries claims = fmap mkQueries gathered
+  where
+    gathered = fmap (\inv -> (inv, store, getBehaviours inv)) invariants
+    invariants = catInvs claims
+    store = head $ (catStores claims) -- TODO: refine AST so we don't need this head anymore
+    getBehaviours (Invariant c _) = filter isPass $ filter (contractMatches c) (catBehvs claims)
+    contractMatches c b = c == (_contract b)
+    isPass b = (_mode b) == Pass
+
+
+-- *** Data *** --
+
+
+data When = Pre | Post
+  deriving (Eq, Show)
+
+newtype Contract = Contract { unContract :: Id }
+  deriving (Show)
+
+newtype Method = Method { unMethod :: Id }
```

Comment:
I'm not sure if I like these as newtypes or just simple type synonym:
```
type Method = Id
```

Then we won't need to bother with these constructors and stuff, it's really just readability

Reply:
Similar to how `Id` is already a synonym for `String`. What do you think?

Reply:
The nice thing is that they ensure you can't e.g. accidentally pass an `Args` instead of a `Store`, so we gain a little typeee safety with them, but I think I agree that they are overkill here....

Reply:
Yeah, I think it's a little overkill

Reply:
imho, this is more about "types as documentation" then giving any extra safety props